### PR TITLE
New style batch_normalization and fixed_batch_normalization

### PR DESCRIPTION
**Pull request by beam2d**
This PR is a part of #3147. It adds double backprop support of `F.batch_normalization` and `F.fixed_batch_normalization`. Currently, the second gradient is written down with plain numpy/cupy APIs without any fusion. It would be better for performance to fuse kernels in CUDA, which is future work.

### CODE REVIEW

**Comment by beam2d** [comment]
Note: this PR largely refactors `batch_normalization.py`: now the function nodes for `F.batch_normalization` and `F.fixed_batch_normalization` are defined by separate classes. I noticed that the behavior with the `config.train` flag would be unintentionally changed; I will investigate this part.

File: chainer/functions/normalization/batch_normalization.py [hunk 1]
```diff
@@ -233,38 +157,296 @@ def backward(self, inputs, grad_outputs):
...
+        ggamma2 = r / gamma
+
+        gx_hat2 = (gggamma2[expander] * gy -
+                   (coeff_m * ggamma1)[expander] * ggx1)
+        gstd2 = -self.inv_std * (r + (x_hat * gx_hat2).sum(axis=self.axis))
+        gmean2 = -self.inv_std * gx_hat2.sum(axis=self.axis)
+        gx2 = self.inv_std[expander] * gx_hat2 + inv_m * (
+            gmean2[expander] + x_hat * gstd2[expander])
+        ggy2 = (gggamma2[expander] * x_hat + ggbeta2[expander] +
+                coeff[expander] * ggx1)
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: How about breaking line before operator `+`? (PEP-8 recommendation)

File: chainer/functions/normalization/batch_normalization.py [hunk 2]
```diff
@@ -233,38 +157,296 @@ def backward(self, inputs, grad_outputs):
...
+
+        gggamma2 = gggamma1 + tmp * gamma_over_var
+        gx_hat = gy * gggamma2[expander]
+        gx2 = self.inv_std[expander] * gx_hat
+        gmean2 = -self.inv_std * gx_hat.sum(axis=self.axis)
+
+        g_gamma_over_std = (ggx1 * gy).sum(axis=self.axis) - ggmean1 * gbeta1
+        ggbeta2 = ggbeta1 - ggmean1 * self.gamma_over_std
+        ggy2 = (gggamma2[expander] * x_hat + ggbeta2[expander] +
+                self.gamma_over_std[expander] * ggx1)
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Ditto. (line continuation)

File: chainer/functions/normalization/batch_normalization.py [hunk 3]
```diff
@@ -233,38 +157,296 @@ def backward(self, inputs, grad_outputs):
...
+
+        ggamma2 = (self.inv_var * g_gamma_over_var +
+                   self.inv_std * g_gamma_over_std)
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Ditto. (line continuation)

File: chainer/functions/normalization/batch_normalization.py [hunk 4]
```diff
@@ -233,38 +157,296 @@ def backward(self, inputs, grad_outputs):
...
+        gvar2 = -(ggamma2 * gamma_over_var + 0.5 * self.inv_var * (
+            (x_hat * gx_hat).sum(axis=self.axis) -
+            self.gamma_over_std * g_gamma_over_std))
```
Comment (niboshi) [inline chainer/functions/normalization/batch_normalization.py]: Ditto. (line continuation)

**Comment by beam2d** [comment]
Fixed.

**Comment by niboshi** [comment]
Please fix test error.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM!
